### ISSUE TITLE
Return type of canonical_unit

### ISSUE THREAD

**Issue by thofma** [Nemocas/Nemo.jl#300]
Currently `u = canonical_unit(r); r = mul!(r, r, u)` does not work generically, since for polynomials the canonical unit is an element of the base ring. We can either
1) Make `canonical_unit(r)` return an element of `base_ring(r)` consistently or
2) Demand that the polynomial interface also includes `mul!(::PolyElem{T}, PolyElem{T}, T)`.

I personally prefer option 1.

**Comment by wbhart** [comment]
Yes, I prefer option 1. We should always explore all possibilities before
adding further unsafe operations.

**Comment by thofma** [comment]
So what should canonical_unit of a m x n matrix return?

**Comment by wbhart** [comment]
Why does that need to be defined?

I can understand why we will need to define it for an nxn matrix algebra,
but not for a generic mxn matrix.

In the case of an nxn matrix algebra, the only logical choice is I_n.

**Comment by wbhart** [comment]
Or possibly canonical_unit(M[1][1])*I_n, though this gets in trouble with
0x0 matrices, where I guess it would have to be defined as the zero matrix.

**Comment by thofma** [comment]
I did not define it for matrices. It is already there as canoncial_unit(M[1, 1]).

I just wanted to know how to do it properly for matrices while I fix it for polynomials.

**Comment by wbhart** [comment]
I could be wrong, but I think it shouldn't be there for matrices. It's only
used for polynomials over a ring, and at present we have no rings of
matrices. It's probably just there in anticipation of an eventual matrix
algebra module.

It certainly shouldn't be canoncial_unit(M[1, 1]), since it is supposed to
be an element in the ring, and this is an element in the base_ring.

**Comment by wbhart** [comment]
... though, I suppose in practice that ends up doing the same thing, since
we just end up dividing by it in various places.

**Comment by wbhart** [comment]
...I think this is done primarily because it doesn't make sense to define
divexact for general matrices. When we finally have matrix algebras, it
will make sense. Then it will probably make more sense (generically
speaking) to define it as a matrix rather than a scalar.

**Comment by thofma** [comment]
OK. This is more complicated than I anticipated. Let us ignore the canonical_unit of a matrix for a moment.

When writing the generic code Hermite normal forms and stuff like that, you usuall want to normalize the diagonal entries. If I have a matrix M and want to multiply row i with the canonical unit of M[i, 1] then I would like to do it as follows:
```
a = canonical_unit(M[i, 1])
ainv = inv(a)
for i in 1:cols(M); mul!(M[i, 1], M[i, 1], ainv); end
```
I have to use mul!, since we don't have inplace division. But with our new convention, a is an element of the base ring, but we don't have inv for all base rings. For example, inv is neither defined for polynomials nor for fmpz. I am tempted to add inv for all rings in Nemo.

**Comment by wbhart** [comment]
I don't think this is a good idea. It will cause problems when we finally
get traits, which will basically boil down to whether or not a ring has an
inv function or not. It would be hard to distinguish a field any other way!

The function divexact makes sense in a ring, since it assumes the division
is exact (even if it does raise an exception if someone uses it
incorrectly).

Of course I see the problem, but I don't think things like this can be
avoided in generic code. If you want the behaviour to be better in some
special rings, you really have to implement something different for those
rings. In this case, you might be able to factor that piece of code out
somehow, so that the amount of duplicated code is small when you implement
special versions for specific rings.

**Comment by wbhart** [comment]
In this case, you might be able to factor out a normalise_diagonal function
(we really must standardise/ze on American or English spelling for all
these function names), which has a special version for fields that uses inv.

**Comment by wbhart** [comment]
To be honest, I'm not sure how long this will last though. Eventually it's
going to be hard to explain to people why they can't do inv for Laurent
series over a ring. It's actually near impossible to keep consistency in a
CAS, which is why there are almost infinitely many bugs, depending on who
is trying to implement what this week.

**Comment by thofma** [comment]
I will go with divexact(one(R), a) then for the moment and I will delete all the inv functions for power series...?

P.S.: Magma has an IsInvertible function, similar to IsDivisbleBy.

**Comment by wbhart** [comment]
I have absolutely no idea what to do about power series. I want to replace
the relative power series with Laurent series, and over a field, these are
invertible, over a ring, they are not. The really big problem is power
series with a unit constant term are invertible.

I'm not in favour of deleting the inv for power series, but I can't explain
why. I think I need some time to think about this.

Maybe traits are going to have to depend on other functions, not thing like
gcd and inv and so on. For example, maybe we'll have to define a function
isfield, which depends only on the type, and make the traits depend on that.

Or perhaps traits will allow us to define a trait function called inv for
fields and a non-trait function called inv, for rings.

Can you define an inv function for all rings? I suppose it can just be
divexact(one(parent(f)), f) as a fallback. But will this actually do what
we want in all cases?

In order to solve the problem you are trying to solve, it might actually be
better to define canonical_unit everywhere to be the inverse of what we
currently define it to be. Then we only have to multiply by it.

Also, are you sure that canonical_unit is really the function you want. It
was only ever designed to be used for canonicalising output for printing of
polynomials. You seem to be putting it to another use here.

**Comment by fredrik-johansson** [comment]
Does inv need to be a total function? If it can throw an exception for any non-invertible input, then it shouldn't be a problem to define it for any ring.

**Comment by wbhart** [comment]
Ok, define inv for every ring if you want. I have concerns both ways, and I
can't foresee which is going to cause more problems.

**Comment by thofma** [comment]
I also wondered if canonical_unit is the function I want. In Poly.jl and Matrix.jl, you use it to get canonical representatives modulo units. Since I want to do the same with diagonal entries of echelon forms, I guess it is the function I want. I will think about it some more.
